server/src/service-layer/controllers/PaymentController.ts

```diff
@@ -0,0 +1,139 @@
+        return { error: "Already a member" }
+      }
+
+      const stripeService = new StripeService()
+      const userDataService = new UserDataService()
+
+      const userData = await userDataService.getUserData(uid)
+
+      /**
+       * Generate customer id if required
+       */
+      let stripeCustomerId: string
+      if (!userData.stripe_id) {
+        const { id } = await stripeService.createNewUser(email, uid)
+        stripeCustomerId = id
+        await userDataService.editUserData(uid, { stripe_id: stripeCustomerId })
+      } else {
+        stripeCustomerId = userData.stripe_id
+        /**
+         * See if user already has active session
+         */
+        const activeSession = await stripeService.getActiveSessionForUser(
+          stripeCustomerId,
+          CheckoutTypeValues.MEMBERSHIP
+        )
+        if (activeSession) {
+          const { client_secret, metadata } = activeSession
+          this.setStatus(200)
+          return {
+            clientSecret: client_secret,
+            membershipType: metadata[
+              MEMBERSHIP_TYPE_KEY
+            ] as MembershipTypeValues,
+            message: "existing session found"
+          }
+        }
+        /**
+         * See if user has pending payment or has recently paid -> if user didn't have a stripe id that means
+         * they couldn't have a completed session
+         */
+        if (
+          (await stripeService.hasRecentlyCompletedCheckoutSession(
+            stripeCustomerId
+          )) ||
+          (await stripeService.hasProcessingPaymentIntent(stripeCustomerId))
+        ) {
+          this.setStatus(409)
+          return {
+            message: "Payment is still being processed"
+          }
+        }
+      }
+
+      /**
+       * Check what price user is going to pay based on the details they filled in
+       */
+      const requiredMembership = new PricingService().getMembershipType(
+        userData
+      )
+
+      /**
+       * Get required product and generate client secret
+       */
+      const requiredMembershipProducts =
+        await stripeService.getProductByMetadata(
+          MEMBERSHIP_TYPE_KEY,
+          requiredMembership
+        )
+
+      // We assume there will only be one active product at a time for memberships (needs to be communicated with admins)
+      const requiredMembershipProduct = requiredMembershipProducts.find(
+        (product) => product.active
+      )
+
+      if (requiredMembershipProduct === undefined) {
+        throw new Error("Product not found")
+      }
+
+      // Note this will throw error if undefined (product wasn't found)
+      const { default_price } = requiredMembershipProduct
+
+      const clientSecret = await stripeService.createCheckoutSession(
+        uid,
+        // Note if the url changes workflows need to be updated to have the deployments work correctly
+        `${process.env.FRONTEND_URL}/register/confirm?session_id={CHECKOUT_SESSION_ID}`,
```

Comment:
Thought this was wrong for a sec, but nope, that's what [stripe wants us to do](https://docs.stripe.com/payments/checkout/custom-success-page)!